zb-db/src/main/java/io/zeebe/db/impl/rocksdb/ZeebeRocksDbFactory.java

```diff
@@ -113,26 +116,132 @@ private ZeebeRocksDbFactory(
+                CompressionType.LZ4_COMPRESSION,
+                CompressionType.LZ4_COMPRESSION))
+        // Target file size for compaction.
+        // Defines the desired SST file size for different levels (but not guaranteed, it is usually
+        // lower)
+        // L0 is what gets merged and flushed, e.g. 3 memtables to X, and target file size and
+        // multiplier is for L1 and other levels.
+        // L1 => 8Mb, L2 => 16Mb, L3 => 32Mb
+        // As levels get bigger, we want to have a good balance between the number of files and the
+        // individual file sizes
+        // https://github.com/facebook/rocksdb/blob/fd0d35d390e212b617e90d7567102d3e5fd1c706/include/rocksdb/advanced_options.h#L417-L429
+        .setTargetFileSizeBase(8 * 1024 * 1024L)
+        .setTargetFileSizeMultiplier(2)
+        // misc
+        .setTableFormatConfig(tableConfig);
 
+    return columnFamilyOptions;
+  }
+
+  private ColumnFamilyOptions createFromUserOptions(
+      final Properties userProvidedColumnFamilyOptions) {
     final var columnFamilyOptions =
-        ColumnFamilyOptions.getColumnFamilyOptionsFromProps(columnFamilyOptionProps);
+        ColumnFamilyOptions.getColumnFamilyOptionsFromProps(userProvidedColumnFamilyOptions);
     if (columnFamilyOptions == null) {
       throw new IllegalStateException(
           String.format(
               "Expected to create column family options for RocksDB, "
                   + "but one or many values are undefined in the context of RocksDB "
-                  + "[Compiled ColumnFamilyOptions: %s; User-provided ColumnFamilyOptions: %s]. "
+                  + "[User-provided ColumnFamilyOptions: %s]. "
                   + "See RocksDB's cf_options.h and options_helper.cc for available keys and values.",
-              columnFamilyOptionProps, userProvidedColumnFamilyOptions));
+              userProvidedColumnFamilyOptions));
     }
     return columnFamilyOptions;
   }
+
+  private DBOptions createDefaultDbOptions(final List<AutoCloseable> closeables) {
+    final var dbOptions =
+        new DBOptions()
+            .setCreateMissingColumnFamilies(true)
+            .setErrorIfExists(false)
+            .setCreateIfMissing(true)
+            .setParanoidChecks(true)
+            // disabling mmap helps improve performance/memory usage when a DB is highly
+            // fragmented with many small files
+            .setAllowMmapReads(false)
+            .setAllowMmapWrites(false)
+            .setMaxOpenFiles(rocksDbConfiguration.getMaxOpenFiles())
+            // do not hog processing if the DB is highly fragmented by opening many threads to
+            // open all files
+            .setMaxFileOpeningThreads(1)
+            // 1 flush, 1 compaction
+            .setMaxBackgroundJobs(2)
+            // may not be necessary when WAL is disabled, but nevertheless recommended to avoid
+            // many small SST files
+            .setAvoidFlushDuringRecovery(true)
```

Comment:
Would also add `setAvoidFlushDuringShutdown` which can also help prevent having too many small SST files (though I'm not sure by how much).